mmf/trainers/lightning_core/loop_callback.py

```diff
@@ -2,35 +2,196 @@
 
 import logging
 
+import torch
+from mmf.common.meter import Meter
 from mmf.common.registry import registry
+from mmf.common.report import Report
+from mmf.trainers.core.reporting import TrainerReportingMixin
+from mmf.utils.logger import calculate_time_left, summarize_report
+from mmf.utils.timer import Timer
 from pytorch_lightning.callbacks.base import Callback
 
 
 logger = logging.getLogger(__name__)
 
 
-class LightningLoopCallback(Callback):
+class LightningLoopCallback(Callback, TrainerReportingMixin):
     def __init__(self, lightning_trainer):
         super().__init__()
         self.lightning_trainer = lightning_trainer
+        self.trainer_config = lightning_trainer.trainer_config
+        self.training_config = lightning_trainer.training
+
+        # for logging
+        self.total_timer = Timer()
+        self.snapshot_iterations = len(lightning_trainer.data_module.val_loader)
+        self.snapshot_iterations //= self.training_config.batch_size
 
     def on_init_start(self, trainer):
         pass
 
     def on_train_start(self, trainer, pl_module):
         registry.register("current_epoch", trainer.current_epoch)
+        self.train_combined_report = None
+
+        self.train_timer = Timer()
+        self.snapshot_timer = Timer()
 
     def on_train_batch_end(
         self, trainer, pl_module, outputs, batch, batch_idx, dataloader_idx
     ):
         # prepare the next batch
         self.lightning_trainer.data_module.train_loader.change_dataloader()
 
+        # aggregate train combined_report
+        self.train_combined_report = self._update_and_create_report(
+            outputs, batch_idx, self.train_combined_report
+        )
+
+        # log
+        if trainer.global_step % self.trainer_config.log_every_n_steps == 0:
+            self._train_log(trainer)
+
+        # eval
+        if trainer.global_step % self.trainer_config.val_check_interval == 0:
+            self._start_eval(trainer)
+
+        # save checkpoints - TODO: @sash
+
     def on_train_end(self, trainer, pl_module):
-        trainer.run_evaluation(test_mode=False)
+        logger.info("Stepping into final validation check")
+        # Only do when run_type has train as it shouldn't happen on validation and
+        # inference runs. Inference will take care of this anyways. Also, don't run
+        # if current iteration is divisble by snapshot interval as it will just
+        # be a repeat
+        if (
+            "train" in self.lightning_trainer.run_type
+            and trainer.global_step % self.trainer_config.val_check_interval != 0
+        ):
+            self._start_eval(trainer)
```

Comment:
i don't quite follow why evaluation needs to be started like this. why isn't the validation loop sufficient?

it seems like this callback is mixing both the logging of event times (could the profiler be used instead?) along with the evaluation hook. imo these should be separated